src/openrct2-ui/windows/Tooltip.cpp

```diff
@@ -26,18 +26,106 @@ static rct_widget window_tooltip_widgets[] = {
     WIDGETS_END,
 };
 
-static void WindowTooltipUpdate(rct_window *w);
-static void WindowTooltipPaint(rct_window *w, rct_drawpixelinfo *dpi);
+// clang-format on
 
-static WindowEventList window_tooltip_events([](auto& events)
+class TooltipWindow final : public Window
 {
-    events.update = &WindowTooltipUpdate;
-    events.paint = &WindowTooltipPaint;
-});
-// clang-format on
+private:
+    utf8 _tooltipText[sizeof(gCommonStringFormatBuffer)]{};
+    int16_t _tooltipNumLines = 1;
 
-static utf8 _tooltipText[sizeof(gCommonStringFormatBuffer)];
-static int16_t _tooltipNumLines;
+public:
+    void SetMessage(const OpenRCT2String& message, ScreenCoordsXY screenCoords)
+    {
+        int32_t textWidth = FormatTextForTooltip(message);
+        int32_t _width = textWidth + 3;
+        int32_t _height = ((_tooltipNumLines + 1) * font_get_line_height(FontStyle::Small)) + 4;
+
+        window_tooltip_widgets[WIDX_BACKGROUND].right = _width;
+        window_tooltip_widgets[WIDX_BACKGROUND].bottom = _height;
+
+        int32_t screenWidth = context_get_width();
+        int32_t screenHeight = context_get_height();
+        screenCoords.x = std::clamp(screenCoords.x - (_width / 2), 0, screenWidth - _width);
+
+        // TODO The cursor size will be relative to the window DPI.
+        //      The amount to offset the y should be adjusted.
+
+        int32_t max_y = screenHeight - _height;
+        screenCoords.y += 26; // Normally, we'd display the tooltip 26 lower
+        if (screenCoords.y > max_y)
+            // If y is too large, the tooltip could be forced below the cursor if we'd just clamped y,
+            // so we'll subtract a bit more
+            screenCoords.y -= _height + 40;
+        screenCoords.y = std::clamp(screenCoords.y, 22, max_y);
+
+        // resize and reposition
+        Invalidate();
+        windowPos = screenCoords;
+        width = _width;
+        height = _height;
+        Invalidate();
```

Comment:
Where did these come from? sound weird. I'm guessing this logic shouldn't be here because the window should've been started with the right `screenCoords` value set, and these initializations maybe should have been left on the `WindowTooltipShow`

Reply:
This is no longer needed since window position and dimension are now calculated in constructor and passed to WindowCreate() function.